### ISSUE TITLE
Rock-Ons do not Support Port mapping for udp AND tcp on the same port

### ISSUE THREAD

**Issue by henfri** [rockstor/rockstor-core#1180]
Hello,

this is my logitechmediaserver/squeezeserver rockon:
{
    "SqueezeBoxServer": {
        "containers": {
            "squeezebox": {
                "image": "tdeckers/logitechmediaserver",
                "launch_order": 1,
                "ports": {
                    "9000": {
                        "description": "Webserver Port. Suggested default: 9000.",
                        "host_default": 9000,
                        "label": "Webserver port",
                        "protocol": "tcp"
                    },
                    "9090": {
                        "description": "CLI Port. Suggested default: 9090.",
                        "host_default": 9090,
                        "label": "CLI port",
                        "protocol": "tcp"
                    },
                    "3483": {
                        "description": "SlimProto Port. Suggested default: 3483.",
                        "host_default": 3483,
                        "label": "Slimproto port",
                        "protocol": "tcp"
                    },
                    "3483": {
                        "description": "SlimProto Port udp. Suggested default: 3483.",
                        "host_default": 3483,
                        "label": "Slimproto port udp",
                        "protocol": "udp"
                    },
                    "9001": {
                        "description": "Supervisord. Suggested default: 9001.",
                        "host_default": 9001,
                        "label": "Supervisord port",
                        "protocol": "tcp"
                    }

                },
                "volumes": {
                    "/media": {
                        "description": "Select the Share containing your Media",
                        "label": "Config Media",
                        "min_size": 1073741824
                    }
                }
            }
        },
        "description": "Server for Squeezebox Devices",
        "ui": {
            "https": true,
            "slug": ""
        },
        "volume_add_support": true,
        "website": "http://mysqueezebox.com",
        "version": "1.0"
    }
}

What's new here is that port 3483 is also mapped as udp, not only tcp.

This does work actually. But in the Web-Interface, only the UDP variant is shown.
I suppose that this is, because the list is built by the Port number and the second occurence overwrites the first while parsing the json.

Proposal is to add the type (udp/tcp) when identifying the ports in the list.

Greetings,
Hendrik

**Comment by henfri** [comment]
Here the output of docker ps:
 docker ps
CONTAINER ID        IMAGE                          COMMAND                  CREATED             STATUS              PORTS                                                                                                                                                                                NAMES
ebfd8b82bd85        tdeckers/logitechmediaserver   "/usr/local/bin/boot."   4 minutes ago       Up 4 minutes        0.0.0.0:9000-9001->9000-9001/tcp, 0.0.0.0:3483->3483/udp, 3483/tcp, 0.0.0.0:9090->9090/tcp  

Actually, I am not sure "3483/tcp" is correct. Shouldn't it be
0.0.0.0:3483->3483/tcp

**Comment by schakrava** [comment]
@henfri , nice observation and thanks for opening this issue. You are right that the second occurrence overwrites the first, but the second is unnecessary. If you like both tcp and udp ports to be open, just don't include "protocol" in the Port object definition. If there is no protocol defined, both tcp and udp will be mapped.

I've updated the readme in the rockon-registry repo. Thanks!

**Comment by henfri** [comment]
Hello Suman,

thanks. 
it does not seem to work this way. The Server cannot discover the client (which runs via 3483)
docker ps shows:
`0.0.0.0:9000-9001->9000-9001/tcp, 0.0.0.0:3483->3483/udp, 3483/tcp, 0.0.0.0:9090->9090/tcp`

I would expect
`0.0.0.0:9000-9001->9000-9001/tcp, 0.0.0.0:3483->3483/udp, 0.0.0.0:3483:3483/tcp, 0.0.0.0:9090->9090/tcp`

In fact, if I start the container with
```
docker run -d \
           -p 9000:9000 \
           -p 9090:9090 \
           -p 3483:3483 \
           -p 3483:3483/udp \
           -v /mnt2/RockOn_Configs_Squeezebox/:/var/lib/squeezeboxserver/prefs \
           -v /var/run/logitechmediaserver:/mnt/state \
           -v /mnt2/Musik:/media \
           tdeckers/logitechmediaserver

```
it works as expected:

> 0.0.0.0:3483->3483/tcp, 0.0.0.0:9000->9000/tcp, 9001/tcp, 0.0.0.0:3483->3483/udp, 0.0.0.0:9090->9090/tcp        

  


Greetings,
Hendrik

**Comment by schakrava** [comment]
Ok, finally I got a chance to test this thoroughly. This is my observation, as expected.

1. If protocol is explicitly provided in the json, then only that protocol is exposed.
2. if protocol is not provided, then both tcp and udp are exposed.
3. I have mariadb and Sickrage, both just added yesterday. mariadb is tcp only whereas, for Sickrage, I've removed protocol entirely for testing. And the output is as expected

$ docker ps
CONTAINER ID        IMAGE                  COMMAND             CREATED              STATUS              PORTS                                            NAMES
35c5c2b4750e        linuxserver/sickrage   "/sbin/my_init"     About a minute ago   Up About a minute   0.0.0.0:8086->8081/udp, 0.0.0.0:8086->8081/tcp   sickrage
53408f412519        linuxserver/mariadb    "/sbin/my_init"     41 hours ago         Up 23 minutes       0.0.0.0:3306->3306/tcp                           linuxserver-mariadb
